File: go/luhn/luhn.go
package luhn

import (
	"strings"
	"unicode"
)

// Valid function validate the luhn number
func Valid(number string) bool {
	number = strings.TrimSpace(number)
	if len(number) < 2 {
		return false
	}
	var sum int
	var dCount int
	numbers := []rune(number)
	for i := len(numbers) - 1; i >= 0; i-- {
		var n = numbers[i]
		if unicode.IsSpace(n) {
			continue
		}
		if !unicode.IsDigit(n) {
			return false
		}
		d := int(n) - '0'
		dCount++
		if dCount%2 == 0 {
			d *= d
			if d > 9 {
				d -= 9
			}
		}
		sum += d
	}
	return sum%10 == 0
}

File: python/luhn/luhn.py
from typing import Tuple


class Luhn:
    def __init__(self, card_num: str):
        self.card_num = None
        self._check_sum = None
        def _dubling(args: Tuple[int, int]) -> int:
            (index, num) = args
            if index % 2:
                num += num
                if num > 9:
                    num -= 9
            return num

        try:
            nums = [ int(num) for num in card_num if not num.isspace() ]
            if len(nums) > 1:
                nums.reverse()
                self._check_sum = sum(map(_dubling, enumerate(nums)))
                self.card_num = nums
        except:
            pass

    def valid(self) -> bool:
        return bool(self.card_num and not self._check_sum % 10)
